examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,892 @@
+
+```rust 
+impl FromSql<crate::schema::smdb::sql_types::ServiceEndpoint, Pg> for Endpoint {
+    fn from_sql(bytes: PgValue<'_>) -> deserialize::Result<Self> {
+        let (name, version, base_uri, port, protocol) =
+            FromSql::<Record<(Text, Integer, Text, Integer, PgProtocolType)>, Pg>::from_sql(bytes)?;
+
+        Ok(Endpoint {
+            name,
+            version,
+            base_uri,
+            port,
+            protocol,
+        })
+    }
+}
+```
+
+Similar to the serialization process, it is paramount that the type annotation match exactly the one used in toSql and
+the type definition in up.sql.
+
+Debugging serialization issues in Diesel is relatively straight forward since the compiler usually points out the
+location of the issue and, often, the issue is a type mismatch that is relatively easy to fix.
+
+It’s worth repeating that you are dealing with three types in total:
+
+1) service_endpoint: The Postgres custom type
+2) ServiceEndpoint: The wrapper struct generated by Diesel that refers to the Postgres type service_endpoint. Also, only
+   the wrapper struct carries the postgres_type annotation in addition to the an sql_type annotation.
+3) Endpoint: The Rust struct with an sql_type annotation referring to the wrapper struct ServiceEndpoint
+
+Make sure all of those types match correctly.
+
+### Rust Struct Service
+
+The service struct gets its serialization and deserialization implementation generated by Diesel so that saves some
+typing. On the other hand, it is a good practice to implement database operations on the actual type itself.
+The wisdom here is twofold. For once, you cannot separate the database operation from the type,
+therefore it makes sense to implement the operations on the type itself. 
+This also hides implementation details in encapsulation,
+as [discussed in the book](https://doc.rust-lang.org/book/ch17-01-what-is-oo.html#encapsulation-that-hides-implementation-details).
+
+Second, you gain a single point of maintenance because, in case your table definition changes, you have to update the
+Rust type anyways and because Diesel is statically checked, the compiler will immediately point out where your
+operations need to be corrected to match the new type definition.
+
+The requirements stated we want CRUD operations, and type-based programming suggest creating a different type per
+operation. Because we use external primary keys, we don’t need an additional type for the delete operation, therefore,
+in total, we crate 3 service types:
+
+1) Service: For Read
+2) CreateService: For create operations
+3) UpdateService: For update operations
+
+The Service and CreateService have identical fields. Also, both types require a primary key annotation in addition to
+the table name. The UpdateService, however, has each field wrapped into an option type. When the option type is Some,
+Diesel knows that this field needs updating. If the field is set to None, Diesel ignores it.
```

Comment:
It might be worth to link the "All about inserts/updates" guides here for the relevant parts?

Reply:
Done.